src/java/org/apache/cassandra/service/reads/ReplicaFilteringProtection.java

```diff
@@ -0,0 +1,473 @@
+ */
+
+package org.apache.cassandra.service.reads;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NavigableSet;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.concurrent.Stage;
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.Columns;
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.DeletionTime;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.ReadCommand;
+import org.apache.cassandra.db.RegularAndStaticColumns;
+import org.apache.cassandra.db.SinglePartitionReadCommand;
+import org.apache.cassandra.db.filter.ClusteringIndexFilter;
+import org.apache.cassandra.db.filter.ClusteringIndexNamesFilter;
+import org.apache.cassandra.db.partitions.UnfilteredPartitionIterator;
+import org.apache.cassandra.db.partitions.UnfilteredPartitionIterators;
+import org.apache.cassandra.db.rows.EncodingStats;
+import org.apache.cassandra.db.rows.RangeTombstoneMarker;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.db.rows.Rows;
+import org.apache.cassandra.db.rows.Unfiltered;
+import org.apache.cassandra.db.rows.UnfilteredRowIterator;
+import org.apache.cassandra.db.rows.UnfilteredRowIterators;
+import org.apache.cassandra.exceptions.ReadTimeoutException;
+import org.apache.cassandra.exceptions.UnavailableException;
+import org.apache.cassandra.locator.Endpoints;
+import org.apache.cassandra.locator.Replica;
+import org.apache.cassandra.locator.ReplicaPlan;
+import org.apache.cassandra.locator.ReplicaPlans;
+import org.apache.cassandra.metrics.TableMetrics;
+import org.apache.cassandra.net.MessagingService;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.StorageProxy;
+import org.apache.cassandra.service.reads.repair.NoopReadRepair;
+import org.apache.cassandra.tracing.Tracing;
+import org.apache.cassandra.utils.btree.BTreeSet;
+
+/**
+ * Helper in charge of collecting additional queries to be done on the coordinator to protect against invalid results
+ * being included due to replica-side filtering (secondary indexes or {@code ALLOW * FILTERING}).
+ * <p>
+ * When using replica-side filtering with CL>ONE, a replica can send a stale result satisfying the filter, while updated
+ * replicas won't send a corresponding tombstone to discard that result during reconciliation. This helper identifies
+ * the rows in a replica response that don't have a corresponding row in other replica responses, and requests them by
+ * primary key to the "silent" replicas in a second fetch round.
+ * <p>
+ * See CASSANDRA-8272 and CASSANDRA-8273 for further details.
+ */
+class ReplicaFilteringProtection<E extends Endpoints<E>>
```

Comment:
I don't think replica filtering protection works with transient replica.. if a node is transient replica, repaired data will be removed on the node purposely to reduce storage requirement and causes unnecessary RFP..

Reply:
@jasonstack Is your worry just that we'll issue useless RFP requests to transient replicas that have already removed their versions of the relevant rows via incremental repair? Is that actually a correctness concern?

It seems like we should be able to ignore transient replicas most of the time anyway as long as we have enough full replicas to meet the original query's read CL...

Reply:
> It seems like we should be able to ignore transient replicas most of the time anyway as long as we have enough full replicas to meet the original query's read CL...

right.. it's not about correctness.. RFP should still work with TR, but have to send dummy protection request to transient replicas even if it's in-sync.

Reply:
Yeah. I don't have any good ideas on how we would avoid that, but this is already a somewhat degenerate case to begin with.

Reply:
I guess we can just skip the requests for transient replicas, though I still have to think a bit more about the possible cases. It's worth mentioning that [secondary indexes are not supported in keyspaces with transient replicas](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/statements/schema/CreateIndexStatement.java#L99-L100), although queries with `ALLOW FILTERING` seem to be supported.

Reply:
@adelapena Let me work through a couple examples in terms of whether or not we would expect problems with filtering queries...

**Example 1**: Say we have 3 replicas (2 full and 1 transient)

1.) `r1` (full) and `r3` (transient) receive a QUORUM write, but `r2` (full) misses it.
2.) `r2` (full) and `r3` (transient) receive a QUORUM delete, but `r1` (full) misses it.
2.) We issue a QUORUM read against `r1` and `r3` (or `r2`). Either way, the stale match from `r1` will be resolved out. _This case looks okay._

**Example 2**: Also 3 replicas (2 full and 1 transient)

1.) Write the value `c=1` at `ONE `with `timestamp=1` to `r1` (full) with `r2` (full) and `r3` (transient) down.
2.) Write the value `c=2` at `ONE` with `timestamp=2` to `r2` (full) with `r1` (full) and `r3` (transient) down.
3.) Write the value `c=3` at `ONE` with `timestamp=3` to `r3` (transient) with `r1` (full) and `r2` (full) down.
4.) Query for `c=1` at `ALL`. The only result filtering on the three replicas should produce is the stale match from `r1`. In this case, it seems like we would need RFP to make sure we get the latest result _from the transient replica_ `r3` to produce the correct result, which is an empty one _until_ `r1` or `r2` is incrementally repaired?

CC @jasonstack

Reply:
@maedhroz So it seems we need to query transient replicas with RFP. Anyway I'm extending the dtests to cover transient replication in case we are missing something.

Reply:
Just added [a couple of very basic dtests](https://github.com/apache/cassandra-dtest/pull/59/commits/c6df6e49be50347467dbf481b9b3176ca7e315c4) using transient replicas. They fail without RFP, and they pass with it.